src/CommitmentsV2/SFA.DAS.CommitmentsV2/Application/Commands/DeleteCohort/DeleteCohortHandler.cs

```diff
@@ -0,0 +1,47 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+using MediatR;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Logging;
+using SFA.DAS.CommitmentsV2.Authentication;
+using SFA.DAS.CommitmentsV2.Data;
+
+namespace SFA.DAS.CommitmentsV2.Application.Commands.DeleteCohort
+{
+    public class DeleteCohortHandler : AsyncRequestHandler<DeleteCohortCommand>
+    {
+        private readonly Lazy<ProviderCommitmentsDbContext> _dbContext;
+        private readonly ILogger<DeleteCohortHandler> _logger;
+        private readonly IAuthenticationService _authenticationService;
+
+        public DeleteCohortHandler(
+            Lazy<ProviderCommitmentsDbContext> dbContext,
+            ILogger<DeleteCohortHandler> logger,
+            IAuthenticationService authenticationService)
+        {
+            _dbContext = dbContext;
+            _logger = logger;
+            _authenticationService = authenticationService;
+        }
+
+        protected override async Task Handle(DeleteCohortCommand command, CancellationToken cancellationToken)
+        {
+            try
+            {
+                var cohort = await _dbContext.Value.Cohorts
```

Comment:
This code is _somewhat_ duplicated from that in the cohort domain service, which does the same basic thing but with some extra checks and specific exceptions:

```
        private static async Task<Cohort> GetCohort(long cohortId, ProviderCommitmentsDbContext db, CancellationToken cancellationToken)
        {
            var cohort = await db.Cohorts.Include(c => c.Apprenticeships).SingleOrDefaultAsync(c => c.Id == cohortId, cancellationToken);
            if (cohort == null) throw new BadRequestException($"Cohort {cohortId} was not found");
            if (cohort.IsApprovedByAllParties) throw new InvalidOperationException($"Cohort {cohortId} is approved by all parties and can't be modified");
            return cohort;
        }
```

Is there some pattern we could adopt to remove the duplication? For example an extension method on the context to fully hydrate the Cohort (eg. include Apprenticeships) and then perform the extra checks to prevent access when it no longer "exists" (deleted, never existed in the first place, or fully approved).

Thoughts?

Reply:
moved into an extension method